common/src/service/ExceptionHandlingHelper.cs

```diff
@@ -116,12 +117,56 @@ public static async Task<ErrorCode> GetExceptionCodeAsync(HttpResponseMessage re
             try
             {
                 IoTHubExceptionResult responseContent = JsonConvert.DeserializeObject<IoTHubExceptionResult>(responseContentStr);
-                Dictionary<string, string> messageFields = JsonConvert.DeserializeObject<Dictionary<string, string>>(responseContent.Message);
 
-                if (messageFields != null
-                    && messageFields.TryGetValue(CommonConstants.ErrorCode, out string errorCodeObj))
+                try
                 {
-                    errorCodeValue = Convert.ToInt32(errorCodeObj, CultureInfo.InvariantCulture);
+                    Dictionary<string, string> messageFields = JsonConvert.DeserializeObject<Dictionary<string, string>>(responseContent.Message);
+
+                    if (messageFields != null
+                        && messageFields.TryGetValue(CommonConstants.ErrorCode, out string errorCodeObj))
+                    {
+                        errorCodeValue = Convert.ToInt32(errorCodeObj, CultureInfo.InvariantCulture);
+                    }
+                }
+                catch (JsonReaderException ex)
+                {
+                    if (Logging.IsEnabled)
+                        Logging.Error(null, $"Failed to deserialize error message into a dictionary: {ex}. Message body: '{responseContentStr}.'");
+
+                    // In some scenarios, the error response string is a ';' delimited string with the service-returned error code.
+                    const char errorFieldsDelimiter = ';';
+                    string[] messageFields = responseContent.Message.Split(errorFieldsDelimiter);
+
+                    if (messageFields != null)
+                    {
+                        _ = messageFields
+                            .Any(field =>
+                            {
+#if NET451 || NET472 || NETSTANDARD2_0
+                                if (field.IndexOf(CommonConstants.ErrorCode, StringComparison.OrdinalIgnoreCase) >= 0)
+#else
+                                if (field.Contains(CommonConstants.ErrorCode, StringComparison.OrdinalIgnoreCase))
```

Comment:
Is contains not available in the other targets? Can we just not use IndexOf for all targets in that case?

Reply:
yeah, contains is available only on ns2.1+, and using indexOf for these newer frameworks generates https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2249, which I was surprised to see

Reply:
That's sad :( These compiler statements looks so messy.